### PR TITLE
Implement Authenticate User

### PR DESCRIPTION
**Pull request by kimlisa**
Task: https://github.com/hwsc-org/hwsc-user-svc/issues/8
- implemented validatePassword and comparePassword in utility.go
- wrote unit test for autheticateUser, and new additions to utility.go
- linted and formatted
- unit tested
screen shot 2019-02-10 at 3 21 34 pm
screen shot 2019-02-10 at 3 21 39 pm

### CODE REVIEW

File: service/service.go [hunk 1]
```diff
@@ -258,6 +258,7 @@ func (s *Service) UpdateUser(ctx context.Context, req *pb.UserRequest) (*pb.User
 	logger.Info("Updated user:", updatedUser.GetUuid(),
 		updatedUser.GetFirstName(), updatedUser.GetLastName())
 
+	updatedUser.Password = ""
```
Comment (faraonc) [inline service/service.go]: Is there where password gets reset to `""`?

File: service/service.go [hunk 2]
```diff
@@ -267,9 +268,64 @@ func (s *Service) UpdateUser(ctx context.Context, req *pb.UserRequest) (*pb.User
...
+		logger.Error(consts.AuthenticateUserTag, consts.ErrInvalidUserEmail.Error())
+		return nil, status.Error(codes.InvalidArgument, consts.ErrInvalidUserEmail.Error())
+	}
+	if err := validatePassword(user.GetPassword()); err != nil {
+		logger.Error(consts.AuthenticateUserTag, consts.ErrInvalidPassword.Error())
+		return nil, status.Error(codes.InvalidArgument, consts.ErrInvalidPassword.Error())
+	}
+
+	lock, _ := uuidMapLocker.LoadOrStore(user.GetUuid(), &sync.RWMutex{})
+	lock.(*sync.RWMutex).Lock()
```
Comment (faraonc) [inline service/service.go]: This should be `RLock()`, because we are only reading the User

File: service/service.go [hunk 3]
```diff
@@ -267,9 +268,64 @@ func (s *Service) UpdateUser(ctx context.Context, req *pb.UserRequest) (*pb.User
...
+	defer lock.(*sync.RWMutex).Unlock()
```
Comment (faraonc) [inline service/service.go]: `RUnlock()`

File: service/service.go [hunk 4]
```diff
@@ -267,9 +268,64 @@ func (s *Service) UpdateUser(ctx context.Context, req *pb.UserRequest) (*pb.User
...
+	}
+
+	if retrievedUser.GetEmail() != user.GetEmail() {
+		logger.Error(consts.AuthenticateUserTag, consts.MsgErrMatchEmail)
+		return nil, status.Error(codes.InvalidArgument, consts.MsgErrMatchEmail)
+	}
+
+	if err := comparePassword(retrievedUser.GetPassword(), user.GetPassword()); err != nil {
+		logger.Error(consts.AuthenticateUserTag, consts.MsgErrMatchPassword, err.Error())
+		return nil, status.Error(codes.Unknown, err.Error())
```
Comment (faraonc) [inline service/service.go]: I think looking up email and password should use `codes.Unauthenticated`

File: service/utility.go [hunk 5]
```diff
@@ -56,6 +56,13 @@ func validateUser(user *pb.User) error {
 	return nil
 }
 
+func validatePassword(password string) error {
+	if password == "" || strings.TrimSpace(password) != password {
```
Comment (faraonc) [inline service/utility.go]: This can be simplify to `if strings.TrimSpace(password) == ""`

**Comment by faraonc** [comment]
I forgot to mention to include a link to the task or epic, so we can reference this PR in case we have to troubleshoot.

File: service/service.go (hunk 1 above)
Comment (kimlisa) [reply service/service.go]: yes

**Pull request merged**